src/t2c.c

```diff
@@ -0,0 +1,302 @@
+ * "LICENSE" for information on usage and redistribution of this file.
+ */
+
+#include <llvm-c/Analysis.h>
+#include <llvm-c/BitWriter.h>
+#include <llvm-c/Core.h>
+#include <llvm-c/ExecutionEngine.h>
+#include <llvm-c/Target.h>
+#include <llvm-c/Transforms/PassBuilder.h>
+#include <stdlib.h>
+
+#include "jit.h"
+#include "riscv_private.h"
+
+#define MAX_BLOCKS 8152
+
+struct LLVM_block_map_entry {
+    uint32_t pc;
+    LLVMBasicBlockRef block;
+};
+
+struct LLVM_block_map {
+    uint32_t count;
+    struct LLVM_block_map_entry map[MAX_BLOCKS];
+};
+
+FORCE_INLINE void t2c_block_map_insert(struct LLVM_block_map *map,
+                                       LLVMBasicBlockRef *entry,
+                                       uint32_t pc)
+{
+    struct LLVM_block_map_entry map_entry = {
+        .block = *entry,
+        .pc = pc,
+    };
+    map->map[map->count++] = map_entry;
+    return;
+}
+
+FORCE_INLINE LLVMBasicBlockRef t2c_block_map_search(struct LLVM_block_map *map,
+                                                    uint32_t pc)
+{
+    for (uint32_t i = 0; i < map->count; i++) {
+        if (map->map[i].pc == pc) {
+            return map->map[i].block;
+        }
+    }
+    return NULL;
+}
+
+#define T2C_OP(inst, code)                                                \
+    static void t2c_##inst(                                               \
+        LLVMBuilderRef *builder UNUSED, LLVMTypeRef *param_types UNUSED,  \
+        LLVMValueRef start UNUSED, LLVMBasicBlockRef *entry UNUSED,       \
+        LLVMBuilderRef *taken_builder UNUSED,                             \
+        LLVMBuilderRef *untaken_builder UNUSED, uint64_t mem_base UNUSED, \
+        rv_insn_t *ir UNUSED)                                             \
+    {                                                                     \
+        code;                                                             \
+    }
+
+#define T2C_BUILD_LLVM_ADDR(reg, rv_member, ir_member)                       \
+    FORCE_INLINE LLVMValueRef t2c_build_##reg##_addr(                        \
+        LLVMValueRef start, LLVMBuilderRef *builder, UNUSED rv_insn_t *ir)   \
+    {                                                                        \
+        LLVMValueRef offset[1] = {LLVMConstInt(                              \
+            LLVMInt32Type(),                                                 \
+            offsetof(riscv_t, rv_member) / sizeof(int) + ir_member, true)};  \
+        return LLVMBuildInBoundsGEP2(*builder, LLVMInt32Type(),              \
+                                     LLVMGetParam(start, 0), offset, 1, ""); \
+    }
+
+T2C_BUILD_LLVM_ADDR(rs1, X, ir->rs1);
+T2C_BUILD_LLVM_ADDR(rs2, X, ir->rs2);
+T2C_BUILD_LLVM_ADDR(rd, X, ir->rd);
+T2C_BUILD_LLVM_ADDR(ra, X, rv_reg_ra);
+T2C_BUILD_LLVM_ADDR(sp, X, rv_reg_sp);
+T2C_BUILD_LLVM_ADDR(PC, PC, 0);
+
+#define T2C_BUILD_LLVM_STORE_IMM32(builder, val, addr) \
```

Comment:
The prefix `T2C_` should be applied to the implementation details, and the macros related to LLVM should have prefix `T2C_LLVM_`.